pkg/region/cache.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2021 Outreach Corporation. All Rights Reserved.
+
+// Description: This file contains the read through cache implementation for
+// this package.
+package region
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	// RegionCachePath is the $HOME/<RegionCachePath> location of the regionpicker cache
+	RegionCachePath = filepath.Join(".outreach", ".cache", "box")
+	// RegionCacheFile is the name of the regionpicker cache file
+	RegionCacheFile = "regions.json"
+)
+
+// cache is the package specific cache instance used. This is a global variable to
+// ensure that this is go-routine safe.
+var cache = &cacheStore{}
+
+// cacheEntry is information on a cloud's cache entry
+type cacheEntry struct {
+	// Duration is how long it took to talk to this region at LastUpdatedAt
+	Duration time.Duration `json:"duration"`
+
+	// ExpiresAt is when this cache entry expires. UTC.
+	ExpiresAt time.Time `json:"expires_at"`
+}
+
+type cacheStore struct {
+	// Clouds is a map of cloudName to cache entries
+	// cloud -> region -> cacheEntry
+	Clouds map[CloudName]map[Name]cacheEntry `json:"clouds"`
+
+	mu sync.RWMutex
+
+	once sync.Once
+}
+
+// expireKeyIfRequired expires a key if it's ready to be expired
+func (c *cacheStore) expireKeyIfRequired(cloud CloudName, r Name, v *cacheEntry) {
+	if time.Now().UTC().After(v.ExpiresAt) {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		delete(c.Clouds[cloud], r)
+	}
+}
+
+// get returns a cache entry for a given cloud/region pairing
+func (c *cacheStore) get(cloud CloudName, r Name) (*cacheEntry, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if _, ok := c.Clouds[cloud]; ok {
+		return nil, false
+	}
```

Comment:
nit:

you shouldn't need this either -- the check on L66 should catch it.  [playground example]
(https://play.golang.org/p/IwaZnRNrvYX)